### PR TITLE
add support for ATTITUDE_QUATERNION

### PR DESCRIPTION
**Pull request by peterbarker**
https://discuss.ardupilot.org/t/can-not-receive-message-attitude-quaternion-using-pymavlink/68219/4

### CODE REVIEW

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -4507,6 +4508,30 @@ void GCS_MAVLINK::send_attitude() const
...
+void GCS_MAVLINK::send_attitude_quaternion() const
+{
+    const AP_AHRS &ahrs = AP::ahrs();
+    Quaternion quat;
+    if (!ahrs.get_secondary_quaternion(quat)) {
+        return;
+    }
+    const float repr_offseq_q[] {0,0,0,0};  // unused, but probably should correspond to the AHRS view?
+    const Vector3f omega = ahrs.get_gyro();
+    gcs().send_text(MAV_SEVERITY_WARNING, "quatin: %f %f %f %f", quat.q1, quat.q2, quat.q3, quat.q4);
```
Comment (rmackay9) [inline libraries/GCS_MAVLink/GCS_Common.cpp:4520]: I guess this is a bit of leftover debug?

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 2]
```diff
@@ -4507,6 +4508,30 @@ void GCS_MAVLINK::send_attitude() const
...
 }
 
+void GCS_MAVLINK::send_attitude_quaternion() const
+{
+    const AP_AHRS &ahrs = AP::ahrs();
+    Quaternion quat;
+    if (!ahrs.get_secondary_quaternion(quat)) {
+        return;
+    }
+    const float repr_offseq_q[] {0,0,0,0};  // unused, but probably should correspond to the AHRS view?
```
Comment (rmackay9) [inline libraries/GCS_MAVLink/GCS_Common.cpp:4518]: I guess we could move this down to line 4531 so we just create and send an empty quaternion and then put a comment beside it saying something like "// unused" or something like that maybe..

**Comment by rmackay9** [comment]
This generally looks ok to me.  I guess this message is only sent in response to a message interval request?  i.e. not sent by default?

I think we are attempting to send a lot of data by default now and probably 80% of it is unused by the most commonly used GCSs.

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: I'm not sure you can create it anonymously inline?

File: libraries/GCS_MAVLink/GCS_Common.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Huh.  Whoops.

Yes, that was just making sure I was getting out the other end what I was putting in.

**Review by peterbarker — COMMENTED** [review]
> This generally looks ok to me. I guess this message is only sent in response to a message interval request? i.e. not sent by default?

Yes, that's correct.  Users might want one of `ATTITUDE` or `ATTITUDE_QUATERNION` - but the former is what we send now and is generally good enough and the latter is a superset of the former's functionality.    You wouldn't want both IMO.  Until we have the latter people can't move to it, of course.

> I think we are attempting to send a lot of data by default now and probably 80% of it is unused > by the most commonly used GCSs.

I agree - which is why I often try to resist putting things into streams controlled by the streamrate parameters.  OTOH, I think the information which GCSs do use is probably all tied up with other stuff:

```
TERRAIN_REPORT
STATUSTEXT
VFR_HUD
RAW_IMU
HWSTATUS
SCALED_PRESSURE2
GLOBAL_POSITION_INT
FILE_TRANSFER_PROTOCOL
SCALED_IMU3
VIBRATION
AHRS
SYSTEM_TIME
SENSOR_OFFSETS
AHRS2
SERVO_OUTPUT_RAW
TIMESYNC
GPS_GLOBAL_ORIGIN
HEARTBEAT
HOME_POSITION
PARAM_VALUE
MISSION_CURRENT
SCALED_IMU2
GPS_RAW_INT
ATTITUDE
SCALED_PRESSURE
REQUEST_DATA_STREAM
SYS_STATUS
POWER_STATUS
BATTERY_STATUS
EKF_STATUS_REPORT
RC_CHANNELS
MEMINFO
NAV_CONTROLLER_OUTPUT
SIMSTATE
```

I've removed the stray comment and moved the definition of the array down a little.

**Review by rmackay9 — APPROVED** [review]

**Pull request merged**